csharp/ql/src/semmle/code/csharp/exprs/Expr.qll

```diff
@@ -283,19 +282,148 @@ class IsTypeExpr extends IsExpr {
+
+  /**
+   * Gets the local variable declaration in this `is` pattern expression, if any.
+   * For example `s` in `x is string { Length: 0 } s`.
+   */
+  LocalVariableDeclExpr getVariableDeclExpr() { result = this.getChild(2) }
+
+  /**
+   * Gets the recursive pattern of this `is` pattern expression. For example,
+   * `{ Length: 5 }` in `x is string { Length: 5 } s`.
+   */
+  RecursivePatternExpr getRecursivePattern() { result = recursivePattern }
+}
+
+/**
+ * A recursive pattern expression, for example `string { Length: 5 } s`
+ */
+class RecursivePatternExpr extends Expr, @recursive_pattern_expr {
+  override string toString() { result = "{ ... }" }
+
+  /**
+   * Gets the position patterns of this recursive pattern, if any.
+   * For example, `(1,_)`.
+   */
+  PositionalPatternExpr getPositionalPatterns() { result = this.getChild(2) }
+
+  /**
+   * Gets the property patterns of this recursive pattern, if any.
+   * For example,
+   * `{ Length: 5 }` in `o is string { Length: 5 } s`
+   */
+  PropertyPatternExpr getPropertyPatterns() { result = this.getChild(3) }
+
+  /**
+   * Gets the type access of this recursive pattern, if any.
+   * For example, `string` in `string { Length: 5 }`
+   */
+  TypeAccess getTypeAccess() { result = this.getChild(1) }
+
+  /**
+   * Gets the local variable declaration of this recursive pattern, if any.
+   * For example, `s` in `string { Length: 5 } s`.
+   */
+  LocalVariableDeclExpr getVariableDeclExpr() { result = this.getChild(0) }
+}
+
+/** A property pattern. For example, `{ Length: 5 }`. */
+class PropertyPatternExpr extends Expr, @property_pattern_expr {
+  override string toString() { result = "{ ... }" }
+
+  /** Gets the `n`th pattern. */
+  Expr getPattern(int n) { result = this.getChild(n) }
+}
+
+/** A property pattern. For example, `(int x, int y)`. */
+class PositionalPatternExpr extends Expr, @positional_pattern_expr {
+  override string toString() { result = "( ... )" }
+
+  /** Gets the `n`th pattern. */
+  Expr getPattern(int n) { result = this.getChild(n) }
+}
+
+/**
+ * A `switch` expression, for example
+ * ```
+ * (a,b) switch {
+ *     (false, false) => true,
+ *     _ => false
+ * };
+ * ```
+ */
+class SwitchExpr extends Expr, @switch_expr {
+  override string toString() { result = "... switch { ... }" }
+
+  /** Gets the expression being switched. For example `x` in `x switch { ... }`. */
+  Expr getExpr() { result = this.getChild(-1) }
+
+  /** Gets the `n`th case of this switch expression. */
+  SwitchCaseExpr getCase(int n) { result = this.getChild(n) }
+}
+
+/** An arm of a switch expression, for example `(false, false) => true`. */
+class SwitchCaseExpr extends Expr, @switch_case_expr {
+  override string toString() { result = "... => ..." }
+
+  /**
+   * Gets the pattern or matched expression in a switch arm, for example `(false, false)` in
+   * `(false, false) => true`.
+   */
+  Expr getPattern() { result = this.getChild(0) }
```

Comment:
Here `Expr` could probably also use the refined type mentioned for `PropertyPatternExpr::getPattern()` above.